clients/src/main/java/org/apache/kafka/common/requests/OffsetFetchResponse.java

```diff
@@ -109,17 +175,48 @@ public OffsetFetchResponse(Struct struct) {
                 int partition = partitionResponse.getInt(PARTITION_KEY_NAME);
                 long offset = partitionResponse.getLong(COMMIT_OFFSET_KEY_NAME);
                 String metadata = partitionResponse.getString(METADATA_KEY_NAME);
-                short errorCode = partitionResponse.getShort(ERROR_CODE_KEY_NAME);
-                PartitionData partitionData = new PartitionData(offset, metadata, errorCode);
-                responseData.put(new TopicPartition(topic, partition), partitionData);
+                Errors error = Errors.forCode(partitionResponse.getShort(ERROR_CODE_KEY_NAME));
+                if (error != Errors.NONE) {
+                    this.partitionErrors.add(error);
+                    if (!PARTITION_ERRORS.contains(error))
+                        topLevelError = error;
+                }
+                PartitionData partitionData = new PartitionData(offset, metadata, error);
+                this.responseData.put(new TopicPartition(topic, partition), partitionData);
             }
         }
+
+        // for version 2 and later use the top level error code (in ERROR_CODE_KEY_NAME) from the response
+        // for older versions there is no top level error in the response and all errors are partition errors,
+        // so if there is a group or coordinator error at the partition level use that as the top level error
```

Comment:
Could we mention that we do this so that the client can depend on the top-level error code regardless of the offset fetch version?